### ISSUE TITLE
subset of matrix should always return the same type (matrix)

### ISSUE THREAD

**Issue by petres** [josdejong/mathjs#2344]
Depending on the index the function subset returns a matrix or a scalar.
How can I set that the return type is always a matrix? Is there another function?

**Comment by josdejong** [comment]
Have you tried the configuration option `predictable`?

**Comment by josdejong** [comment]
I just tried it but I see `predictable` is not implemented there.

so, what we would like is:

```js
math.config({ predictable: false })
math.subset(math.matrix([1,2,3,4]), math.index(1)) // number 2

math.config({ predictable: true })
math.subset(math.matrix([1,2,3,4]), math.index(1)) // Matrix/Array containing one value 2
```


Anyone interested in making `subset` adhere to the `predictable` config and always return a matrix in that case? Help would be welcome

**Comment by petres** [comment]
`subset` should always return a `Matrix`. If the user wants to extract an element of that matrix he could use `get`. In the case he doesn't know about the returned dimensionality (because the param index is dynamic) she will always prefer to get the same type.

**Comment by josdejong** [comment]
What "the user" wants depends on the use case and which user you're talking about 😄 . 

In a programmatic JavaScript environment it indeed makes sense to use two different, explicit methods `get` and `subset`, each giving back predictable results. 

The method `subset` (and the whole of mathjs) was originally created though for end users using the expression parser to do mathematics. It focuses on making it easy to work with matrices, and was heavily inspired by how you can work with matrices in mathematical applications like Matlab, Octave, Maple, Mathematica. 

So in the expression parser you do things like the following, where it is implicit whether you get back a value or a matrix:

```
A = [1, 2; 3, 4]
myvalue = A[1, 2]    # will return number 2
mysubset = A[2, 1:2] # will return the second row [[3, 4]]
```
You could argue to make an explicit API in the expression parser too, but this is how the mathematical world is used to work and it works like a charm.

Similarly, a function like `sqrt` will normally return a complex result instead of a number with value NaN when you enter a negative value, like `sqrt(-4)`. That makes  the function unpredictable. In some cases like mathematical end users this is desirable, in other cases like programmatically using the function you want a predictable output type. That's why the configuration option `predictable` was introduced, and I think this issue will be addressed by extending `subset` to adhere to this config option.

**Comment by dvd101x** [comment]
By checking other references I think the logic is more in line of the type of the input instead of the size of the index so that if every range in index comes from a scalar then it yields a scalar.
```jl
A = [1, 2; 3, 4]
A[1, 2]      # yields  2
A[1, [2]]    # yields [2]
A[[1], 2]    # yields [2]
A[1:1, 2]    # yields [2]
A[1, 2:2]    # yields [2]
A[2, 1:2]    # yields [3, 4]
A[[2], 1:2]  # yields [[3, 4]]
```
I see also some differences on higher dimension arrays regarding if it comes from an array or comes from a range.

Some of that shows some promising capabilities.

**Comment by gwhitney** [comment]
This issue came up in #2994, and I think there is a clear convention that could/should be adopted here that will eliminate any use of the "predictable" mechanism. That is to say, I would propose that indexing any dimension with a scalar eliminates that dimension, whereas indexing it with an array leaves that dimension in place, regardless of the length of the array. For example, if `A =[1,2,3; 4,5,6]` then (with 1-based indexing as in the parser):
```
A[2,3]  --> 6
A[[1,2],[2,3]] --> [2,3; 5,6]
A[[2],[2,3]] --> [[5, 6]]
A[2,[2,3]] --> [5,6]
A[[1,2],[2]] --> [2; 5]  // which is the same as [[2], [5]]
A[[1,2],2] --> [2,5]
A[[2], [2]] --> [[5]]
A[2, [2]] --> [5]
A[[2], 2] --> [5]
```
This way, the _type_ of the result is exactly predictable just from the types of the indices; it's compatible with boolean indexing as recently implemented; and I believe it covers the original poster's use-case. Thoughts? The change from current behavior is that for example all three of the last expressions currently return just 5. Conceptually, it would not be the _count_ of entries that determines the type of the result, but just the types of the indices.

There are some edge cases with empty vectors for indices:
```
A[1,[]] --> []  // Seems clear
A[[],2] --> []  // Seems OK, no rows.
A[[],[]] --> ??? // Seems like it should be 2-D, but it also seems like it should have no rows. Is there any distinction
// between a 1-D and 2-D empty matrix in mathjs? Right now, I don't think so... There is [[]], but that seems to have one (empty) row, rather than being 2D but having no rows...
```
but I think these things could be ironed out in some reasonable way, I don't think they detract from the basic proposal.

**Comment by josdejong** [comment]
Ow, that is a great idea! I love it. It definitely makes sense to me.

**Comment by dvd101x** [comment]
It also makes sense to me.

As a reference for the edge cases, in numpy it would be like this:
```jl
A[[],[]] --> []
```

The same would apply for N dimensions:
```jl
A[[],2,3]   --> []
A[1,[],3]   --> []
A[1,2,[]]   --> []
A[1,[],[]]  --> []
A[[],2,[]]  --> []
A[[],[],3]  --> []
A[[],[],[]] --> []
```

**Comment by gwhitney** [comment]
This is actually a duplicate of the much older #1484, but since the discussion here is more in-depth, I'm actually going to close the older one.

**Comment by DbCrWk** [comment]
> This is actually a duplicate of the much older #1484, but since the discussion here is more in-depth, I'm actually going to close the older one.

I originally opened #1484, but didn’t get around to opening a PR. This resolution looks great!

**Comment by josdejong** [comment]
👍 thanks David!

**Comment by herdrick** [comment]
I'm not sure if this is the correct place or time to discuss this, but for some matrix M, ex. `M = [[1, 2], [3, 4]]`, if you set a slice to a vector:

`
M[1, :] = [11, 22]
`

... then for my part, I would think that querying that slice should give the same vector. But instead it gives:

`M[1, :] -> [[11, 22]]`

**Comment by herdrick** [comment]
A second thought is that, again for a matrix `M = [[1, 2], [3, 4]]`, since `M[1]` currently throws a DimensionError, why not specify that `M[1] -> [1, 2]` ?  It looks like maybe this wouldn't collide with anything, but I haven't carefully looked in to it. It would conform with the semantics of Numpy and MATLAB, and perhaps others, idk.

**Comment by gwhitney** [comment]
> I'm not sure if this is the correct place or time to discuss this, but the symmetry of getting and setting seems like a good goal. For example, for some matrix M, ex. `M = [[1, 2], [3, 4]]`, if you set a slice to a vector:
> 
> `M[1, :] = [11, 22]`
> 
> ... then for my part, I would think that querying that slice should give the same vector. But instead it gives:
> 
> `M[1, :]` --> `[[11, 22]]`
> 
> Failing that, if the project is committed to `M[1, :]` --> `[[11, 22]]`, then perhaps the setting should be `M[1, :] = [[11, 22]]`

Sorry, I'm not following; I think the conclusion is that if `A=[[1, 2], [3, 4]]` then `A[1, :]` should be `[1, 2]`, not `[[1, 2]]` as is currently returned. It just isn't implemented yet.

As to your point about get-set symmetry, it's made murky by the fact that the source of the assignment is replicated as needed to fill the destination: After `A=[1,2; 3,4; 5;6]; A[[2,3],:] = [7, 8]` we have that A is `[[1, 2], [7, 8], [7, 8]]`. So the 1-dimensional row has been replicated to fill the 2-dimensional "hole" of the specified subset. I don't see much reason to abandon this setting convention, but it does mean that what can appear on the RHS of an assignment is pretty flexible, whereas of course getting a subset will just return one specific shape. Thus, with the mathjs conventions, we can't really rely on get-set symmetry as a guide.

I also don't see a lot of motivation to allow `M[1]` when `M` is two-dimensional; it's reasonably likely to be a typo and to me it seems ambiguous between meaning `M[1, :]` and `M[:, 1]`. Not clear why one should be privileged over the other, especially when both are so convenient to write.

**Comment by herdrick** [comment]
>  I think the conclusion is that if A=[[1, 2], [3, 4]] then A[1, :] should be [1, 2], not [[1, 2]] as is currently returned. It just isn't implemented yet.

Oh I'm sorry, I didn't realize that was the conclusion. So that takes care of my concern about get-set symmetry because the new getting functionality will line up with the setting in that simpler sort of case I mentioned, and also negates the utility of my suggestion about M[1]! 👍

**Comment by josdejong** [comment]
I guess `math.Index` normalizes scalar inputs into arrays, so it always only has to deal with arrays. If we now need to know whether the index originally had a scalar value or not, we have to make changes. At first sight, I think Option 2 would be most neat, but I'm not sure how much impact it will have on other methods.

**Comment by gwhitney** [comment]
I agree, the internal representation of an Index should be totally irrelevant to the outside world, so you should have a free hand to use whatever representation is best suited to the task at hand. If immutability is important, you can make Index objects immutable in any of a few different ways as appropriate: by having its interface entirely consist of getters, by Object.freeze()ing them before you give them back, or by returning Proxy objects that trap and deny any `set` actions. There may even be other ways.

**Comment by dvd101x** [comment]
Thanks Jos and Glen for your insights. Will take them into account in the next steps.

**Comment by dvd101x** [comment]
Hi, just to report some progress

Option 2 seems to work.

https://github.com/josdejong/mathjs/tree/matrix-subset-according-to-type-of-input

Still haven't changed the behaviour of the subset, but changed the index so that it maintains the scalar and doesn't convert to a `matrix`

**Comment by josdejong** [comment]
Thanks for the update David, this looks promising already 👍.

**Comment by josdejong** [comment]
Implemented via #3485.

**Comment by josdejong** [comment]
Published now in `v15.0.0`.
